### PR TITLE
Add JS support for Chrome on Android + Authenticator App

### PR DESCRIPTION
**Pull request by ngm0**

### CODE REVIEW

File: u2f-gae-demo/war/js/u2f-api.js [hunk 1]
```diff
@@ -168,6 +180,17 @@ u2f.getChromeRuntimePort_ = function(callback) {
 };
 
 /**
+ * Return a 'port' abstraction to the Authenticator app.
+ * @param {function(u2f.WrappedChromeRuntimePort_)} callback
+ * @private
+ */
+u2f.getAuthenticatorPort_ = function(callback) {
+  setTimeout(function() {
```
Comment (leshi) [inline u2f-gae-demo/war/js/u2f-api.js]: why use setTimeout instead of invoking the callback?

File: u2f-gae-demo/war/js/u2f-api.js [hunk 2]
```diff
@@ -205,6 +264,244 @@ u2f.WrappedChromeRuntimePort_.prototype.addEventListener =
...
+    function(eventName, handler) {
+  var name = eventName.toLowerCase();
+  if (name == 'message' || name == 'onmessage') {
+    var self = this;
+    /* Register a callback to that executes when
+     * chrome injects the response. */
+    window.addEventListener(
+        'message', self.onRequestUpdate_.bind(self, handler), false);
+  } else {
+    console.error('WrappedAuthenticatorPort only supports onMessage');
```
Comment (leshi) [inline u2f-gae-demo/war/js/u2f-api.js]: in this case, why don't you have to addEventListener for onmessage?

File: u2f-gae-demo/war/js/u2f-api.js [hunk 3]
```diff
@@ -205,6 +264,244 @@ u2f.WrappedChromeRuntimePort_.prototype.addEventListener =
...
+ * Fixup the response provided by the Authenticator to conform with
+ * the U2F spec.
+ * @param {Object} responseData
+ * @return {Object} the U2F compliant response object
+ */
+u2f.WrappedAuthenticatorPort_.prototype.doResponseFixups_ =
+    function(responseObject) {
+  if (responseObject.hasOwnProperty('responseData')) {
+    return responseObject;
+  } else if (this.requestObject_['type'] != u2f.MessageTypes.U2F_SIGN_REQUEST) {
```
Comment (leshi) [inline u2f-gae-demo/war/js/u2f-api.js]: maybe add a case for else to log an error?

File: u2f-gae-demo/war/js/u2f-api.js (hunk 1 above)
Comment (ngm0) [inline u2f-gae-demo/war/js/u2f-api.js]: It's just convention (see the constructor for the Chrome port), the idea being that if the constructor is changed in the future to be blocking, then the callers, like the one here, will just work.

Though in this case, the page's JS should have already used setTimeout.

File: u2f-gae-demo/war/js/u2f-api.js (hunk 2 above)
Comment (ngm0) [inline u2f-gae-demo/war/js/u2f-api.js]: Actually only handles 'message'.  Fixed.

File: u2f-gae-demo/war/js/u2f-api.js (hunk 3 above)
Comment (ngm0) [inline u2f-gae-demo/war/js/u2f-api.js]: else is the success case (fall-through) , it's only sign responses that need fixup.  Added a comment to the effect.

**Comment by leshi** [comment]
When I run the patched-in code and try to register a USB device, I get the following error in the CryptoToken extension:

```
Error in event handler for (unknown): TypeError: Cannot read property 'type' of null
    at handleWebPageRequest (chrome-extension://kmendfapggjehodndflmmgagdbamhnfd/webrequest.js:110:18)
    at messageHandler (chrome-extension://kmendfapggjehodndflmmgagdbamhnfd/cryptotokenbackground.js:109:19)
    at chrome-extension://kmendfapggjehodndflmmgagdbamhnfd/cryptotokenbackground.js:141:17
```

Tracing back, it looks like u2f-api.js (u2f.register) is calling ```postMesasge()``` with ```null```.  Tracing back a bit more, I think ```u2f.WrappedChromeRuntimePort_.prototype.formatRegisterRequest_()``` is returning ```null```.  I think that's happening because the constructed ```req``` object is not being returned.

**Comment by ngm0** [comment]
Ah, nice catch.  Fixed.

**Pull request merged**